### ISSUE TITLE
Migrating to the in-kernel KVDO, or "Recovery journal is in the old format"

### ISSUE THREAD

**Issue by thememika** [dm-vdo/kvdo#85]
After reading the 6.9-rc1 update notice from Linus, I was excited to know that KVDO has been finally merged to the Linux kernel. 
But unfortunately, I'm not able to use any of my production VDO devices with it. The recovery journal is in old format, and as I got, there is no way to bring the devices up for r/w without re-creating them and copying terabytes of data.
```
[ 5813.749275] device-mapper: vdo: dm-vdo3:journal: Recovery journal is in the old format, a read-only rebuild is required.: VDO Status: Unsupported component version (1471)
```
I don't understand why you change the format of journal without writing a conversion code for it. What is the reason behind doing it?
I can't use the in-kernel KVDO because of that.
Is it true that everyone will now have to re-create the VDO devices from scratch, and copy all the data over?

**Comment by rhawalsh** [comment]
Hi @thememika, can you share which version you were using prior to upgrading to the in-kernel version?  I'm glad you were able to discover the `forced rebuild` operation to get back into operation.

I'd like for us to reproduce this situation so we can better understand it and figure out what needs to be done.

Thanks,
-Andy

**Comment by rhawalsh** [comment]
Hi @thememika.  We'll take a look at it.  Thank you for the report!

Just to clarify, you did the forced rebuild on the old (8.2.3.3) version, or the in-kernel 6.9-rc2 version?

**Comment by rhawalsh** [comment]
Hi @thememika, just wanted to give a quick update.

I ran through a few scenarios, and I was able to reproduce what you're seeing.  I'm still playing around with things to figure out how best to document this.  However, the bottom line is to try and always make sure you're cleanly shutting down the VDO volume(s) before making any changes.  Though in my testing from the 8.2.3.3 to 6.9-rc2 on Fedora Rawhide, I was able to repair the volume going both to and from either end after a dirty shutdown by using the `vdoforcerebuild` utility as mentioned.

My reproduction environment was done using an iSCSI target and two initiators.
  - Target Config: RHEL8 with 150G backing storage
  - Initiator 1: RHEL9 with vdo, vdo-support, kmod-kvdo, and lvm2 installed
  - Initiator 2: Fedora Rawhide with vdo, vdo-support, and lvm2 using 6.9-rc2 kernel.

I went through a few scenarios.
1. Graceful shutdown and transfer from RHEL9->Rawhide.
2. Graceful shutdown and transfer from Rawhide->RHEL9.
3. Unclean shutdown (using `echo b > /proc/sysrq-trigger`) on RHEL9 followed by transfer to Rawhide.
4. Unclean shutdown (using `echo b > /proc/sysrq-trigger`) on Rawhide followed by transfer to RHEL9.

Scenario 1:
  - Export a LUN from the target to Initiator 1, and Initiator 2
  - Log into iSCSI target from Initiator 1
  - Create a VDO volume (`vgcreate vg_name /dev/sda; lvcreate -l 100%FREE -V 1T -n vdo_name vg_name`)
  - Format that volume with XFS (`mkfs.xfs -K /dev/vg_name/vdo_name`) and mount it (`mount /mnt/vdo`)
  - Copy some data to it (`cp -a /usr /mnt/vdo`)
  - Gracefully stop the VDO volume (`lvchange -an /dev/vg_name/vdo_name`)
  - Log out of iSCSI session on Initiator 1
  - Log into iSCSI target from Initiator 2
  - Verify that VDO volume is recognized and activated (via `lsblk`)
  - Mount VDO volume and verify the contents.
Everything in Scenario 1 worked as expected.

Scenario 2 is the same as Scenario 1, but with the initiators swapped.

Scenario 3 replaced the step where VDO is gracefully stopped with a forced reboot via `echo b > /proc/sysrq-trigger` followed by a removal of the initiator's access to the LUN on the target to prevent a possible multiple activation situation.  Then, Initiator 2 logged into the target, where we could see the same errors you reported previously.  Upon running `vdoforcerebuild /dev/mapper/vg_name-vpool0_vdata` followed by a deactivate/activate cycle (`lvchange -an /dev/vg_name/vdo_name; lvchange -ay /dev/vg_name/vdo_name`).  At this point the volume's normal operation has resumed after performing the read-only rebuild.  The remaining steps in Scenario 1 and 2 continued from here.

Scenario 4 is the same as Scenario 3, but with Initiators swapped, once again.

**Comment by rhawalsh** [comment]
I did a little bit more digging into the behavior that was experienced here, since just a basic 'unclean shutdown' shouldn't really render a VDO volume to require a forced rebuild.

If I take the same setup as mentioned above, and set up a volume on a RHEL9 host, forcibly reset the host, and then let the system come back up, the VDO volume starts with an automatic recovery and no manual intervention required.  It is specifically the act of trying to move the dirty volume from RHEL9 to the upstream version that causes this particular behavior.

I don't believe this is particularly a new thing with VDO volumes.  We've always had an "upgrade" when moving from one version to the next.  Though, I can't say that I've seen the issue where simply attempting to start a dirty volume on the new version would cause it to require a force-rebuild.  That's something I need to talk with the team about a bit more.

**Comment by lorelei-sakai** [comment]
That is correct. vdoforcerebuild, and most of the other userspace tools, must operate directly on the pool_data device.

Thanks for your kind words, and I'm glad it's working out for you.
